zipline/pipeline/loaders/events.py

```diff
@@ -0,0 +1,183 @@
+import abc
+import pandas as pd
+from six import iteritems
+from toolz import merge
+
+from .base import PipelineLoader
+from .frame import DataFrameLoader
+from .utils import next_date_frame, previous_date_frame, previous_value
+
+WRONG_COLS_ERROR = "Expected columns %s for sid %s but got columns %s."
+
+BAD_DATA_FORMAT_ERROR = ("Data for sid %s must be in DataFrame, "
+                         "Series, or DatetimeIndex.")
+
+SERIES_NO_DTINDEX_ERROR = ("Got Series for sid %d, but index was not "
+                           "DatetimeIndex.")
+
+DTINDEX_NOT_INFER_TS_ERROR = ("Got DatetimeIndex for sid %d.\n"
+                              "Pass `infer_timestamps=True` to use the first "
+                              "date in `all_dates` as implicit timestamp.")
+
+DF_NO_TS_NOT_INFER_TS_ERROR = ("Got DataFrame without a '%r' column for sid "
+                               "%d.\nPass `infer_timestamps=True` to use the "
+                               "first date in `all_dates` as implicit "
+                               "timestamp.")
+
+TS_FIELD_NAME = "timestamp"
+SID_FIELD_NAME = "sid"
+
+
+class EventsLoader(PipelineLoader):
+    """
+    Abstract loader.
+
+    Does not currently support adjustments to the dates of known events.
+
+    Parameters
+    ----------
+    all_dates : pd.DatetimeIndex
+        Index of dates for which we can serve queries.
+    events_by_sid : dict[int -> pd.DataFrame], dict[int -> pd.Series],
+    or dict[int -> pd.DatetimeIndex]
```

Comment:
this is more `dict[int -> pd.DataFrame or pd.Series or pd.DatetimeIndex]` because it does not need to be homogenous.